File: cyclic_rotation/cyclic_rotation.py
# https://app.codility.com/demo/results/training45MYT9-23W/


def get_rotated_index(rotation):
    return rotation if rotation < length else rotation - length


def solution(A, K):
    length = len(A)

    if not length or K == 0 or K % length == 0:
        return A

    rotation = K % length
    result = A.copy()

    for i in range(0, length):
        result[get_rotated_index(rotation + i)] = A[i]

    return result

File: cyclic_rotation/cyclic_rotation.js
function getRotatedIndex(rotation, length) {
  return rotation >= length ? rotation - length : rotation;
}

function solution(A, K) {
  // write your code in JavaScript (Node.js 8.9.4)
  const length = A.length;

  if (!length || K == 0 || K % length) {
    return A;
  }

  let rotation = K % length;

  const result = A.slice();

  for (let i = 0; i < length; i++) {
    result[getRotatedIndex(rotation + i)] = A[i];
  }

  return result;
}
